lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php

```diff
@@ -180,7 +180,14 @@ public function connect($connectionName = null)
                 $this->connections['slave'] = $this->connections['master'];
             }
         } else {
-            $this->connections['slave'] = $this->_conn = $this->connectTo($connectionName);
+            try {
+                $this->connections['slave'] = $this->_conn = $this->connectTo($connectionName);
+            } catch(\Exception $ex) {
+                //failover when slave is dead
+                $this->connections['slave'] = $this->_conn =
```

Comment:
This scenario still requires a test. Also, I think it's really time for us to have a proper connection pooling abstraction for 3.x :-\